tests/flytekit/unit/core/test_dataclass.py

```diff
@@ -27,3 +28,79 @@ def wf() -> AppParams:
 
     res = wf()
     assert res.region == "us-west-3"
+
+
+def test_dataclass_no_json():
+    @dataclass
+    class AppParams(object):
```

Comment:
Could we add more type here? like structuredDataset and FlyteDirectory.

Reply:
and Any or Pickle type

Reply:
I added `StructuredDataset` and `FlyteDirectory` types to the complex type test. Can you clarify what you mean by Any/Pickle types?

Reply:
I mean `typing.List[Any]`. https://docs.flyte.org/projects/cookbook/en/latest/auto/core/type_system/flyte_pickle.html

Reply:
I get `ValueError: Type of Generic List type is not supported, Type typing.Any not supported currently in Flytekit. Please register a new transformer`

Is this supposed to work? I don't see registration of a `TypeTransformer` for `typing.Any`. I see the `PickleTransformer` but it doesn't seem to register for `typing.Any`.

Reply:
Ok I've dug in a little more. Some thoughts:

I copied how `test_type_engine.py` tests the pickle transformer, and it works fine. But it is worth noting that for some reason resolution of something as a `FlytePickle` type, happens in `interface.py` and not in `TypeEngine`, which I think is leading to some weird behavior unexpected behavior, due to the interface only following along the generic type tree, rather than recursing via `TypeEngine.get_transformer`. 
Some examples:
* `List[Foo]` works in a function signature
* `List[Foo]` doesn't work in a dataclass
* `Foo` works both in a function signature and a dataclass
* `A dataclass containing a dataclass field `Foo` works in neither a function signature nor a dataclass.
* A custom `MyGeneric[Foo]` where the user  creates a `TypeTransformer` registered for `MyGeneric` will work in neither.

Rather than replacing the type, I would just "fall back" to `PickleTransformer` inside of `TypeEngine.get_transformer` (with a warning ofc). 
* Gets rid of the `FlytePickle` type completely.
* `get_transformer` could no longer raise a `TypeTransformerFailedError` killing a ton of code (67 usages in flytekit).
Instead e.g. `List[Any]` would fall back to pickle for the list type.

I'm not sure if this level of permissiveness is desirable. That said, I don't think it's any more permissive than the current state of things, it's just less arbitrary. (i.e. I couldn't accurately tell you right now in what situations things will "just work" using pickle, and in what situations they won't, so in that sense, having it "just work" in every situation seems better). 

WDYT. I'm happy to bundle that up in this PR, or I can split it out into a separate PR on top of this.

I have a hunch there's a few more inconsistencies between `interface.py` and the way this PR is doing things, but I haven't explored further

ALTERNATIVE APPROACH

My personal feeling is that pickle should be entirely explicit (i.e. just `TypeEngine.register(PickleTransformer(), additional_types=[Foo])` isn't that hard), but if we wanted to leave the current inconsistent approach in place, we could at least add an `Annotation` based escape hatch that doesn't interact with the type registry at all.

```
from flytekit import PickleTransformer

class Foo:
    ...

PickledFoo = Annotated[Foo, PickleTransformer()]

@task
def my_task(foo: PickledFoo) -> PickledFoo:
    ...
```

then in `TypeEngine.get_transformer` when you test for annotation, check for an explicit transformer, and short circuit the search in the registry.

This would get me out a lot of sticky situations with the registry in fact. 

For example when you're trying to define a second more custom transformer for a type that's already defined, and you only want to use some of the time, you could provide a type alias that explicitly uses that transformer.

Reply:
> My personal feeling is that pickle should be entirely explicit

we can't do that now. it will break people, but can definitely add a hook like you said to override default transformer in flytekit

one of the reason we have pickleType is that we save [original_python_type](https://github.com/flyteorg/flytekit/blob/87c665becca35567063472bcf8477e0d3876f088/flytekit/types/pickle/pickle.py#L36) in the pickle type, and it allows us to check if the type matches between two tasks.

For example, below wf should fail to be registered.
```python
@task
def t1() -> Foo:
  ...

@task
def t2(b: Bar):
  ...

@workflow
def wf():
  t2(b=t1())
```

Reply:
yes, I think we can remove `interface._change_unrecognized_type_to_pickle ` and `FlytePickle`, but we should keep `pickle format` and pickleTransformer.
If get_transformer() fail to find the transformer for the type, flytekit falls back to `PickleTransformer`

Reply:
we can add origin_python_type to literalType [here](https://github.com/flyteorg/flytekit/blob/87c665becca35567063472bcf8477e0d3876f088/flytekit/types/pickle/pickle.py#L92) instead

Reply:
Sounds good. I took a shot at this if you want to take a look. 

I think the Annotated override is probably worth doing, but seems isolated enough it can probably go in a separate PR